common/metrics/grpc.go

```diff
@@ -0,0 +1,110 @@
+package metrics
+
+import (
+	"context"
+	"encoding/json"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+
+var metricsContextMd = "metrics_context_md"
+
+func ClientMetricsTrailerPropagatorInterceptor(
+	ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	var md metadata.MD
+	optsNew := append(opts, grpc.Trailer(&md))
+	err := invoker(ctx, method, req, reply, cc, optsNew...)
+
+	metricUpdates := md.Get(metricsContextMd)
+	if metricUpdates == nil {
+		return err
+	}
+
+	for _, str := range(metricUpdates) {
+		data := ([]byte)(str)
+		var metricValues map[string]int64
+		error := json.Unmarshal(data, &metricValues)
+		if error != nil {
+			continue
```

Comment:
log an error?

Reply:
I'm a bit skeptical about this in general because it most likely would spam a log message on every rpc call to any server service. It would be dangerous for both: logging system and latency. Maybe log every N'th invocation or every N seconds.

Also, I'm using json here, but as we discussed, this most likely would change to protobuf or another protocol.